packages/Button/src/Button.tsx

```diff
@@ -267,61 +281,124 @@ const baseStyle = css`
+  children?: React.ReactNode;
+  disabled?: boolean;
+}
 
-  static propTypes = {
-    variant: PropTypes.oneOf(['default', 'primary', 'info', 'danger', 'dark']),
-    size: PropTypes.oneOf(['xsmall', 'small', 'normal', 'large']),
-    className: PropTypes.string,
-    children: PropTypes.node,
-    disabled: PropTypes.bool,
-    as: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
-    href: PropTypes.string,
-  };
+interface LinkButtonProps extends HTMLElementProps<'a'>, SharedButtonProps {
+  href: string;
+}
+
+interface ButtonButtonProps
+  extends HTMLElementProps<'button'>,
+    SharedButtonProps {
+  href?: null;
+}
+
+type CustomElementButtonProps = SharedButtonProps & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type ButtonProps =
+  | LinkButtonProps
+  | ButtonButtonProps
+  | CustomElementButtonProps;
+
+function hasCustomRoot(props: ButtonProps): props is CustomElementButtonProps {
+  return (props as any).as != null;
+}
+
+function hasLinkRoot(
+  props: LinkButtonProps | ButtonButtonProps,
+): props is LinkButtonProps {
+  return props.href != null;
+}
+
+const omitProps = [
+  'as',
+  'className',
+  'disabled',
+  'size',
+  'variant',
+  'children',
+] as const;
+
+export default class Button extends Component<ButtonProps> {
+  static displayName = 'Button';
 
   static defaultProps = {
     variant: 'default',
     size: 'normal',
     className: '',
     children: null,
     disabled: false,
-    type: 'button',
   };
 
   render() {
-    const {
-      className,
-      children,
+    const { className, children, disabled, variant, size } = this.props;
+
+    const commonProps = {
+      className: cx(
+        baseStyle,
+        buttonSizes[size],
+        buttonVariants[variant],
+        className,
+      ),
       disabled,
-      variant,
-      size,
-      as,
-      href,
-      ...rest
-    } = this.props;
+      'aria-disabled': disabled,
+    };
 
-    let Root = href ? 'a' : 'button';
+    if (hasCustomRoot(this.props)) {
+      const Root = this.props.as;
+      const rest = omit(this.props, omitProps);
 
-    if (as) {
-      Root = as;
+      return (
+        <Root {...rest} {...commonProps}>
+          {children}
+        </Root>
+      );
     }
 
+    if (hasLinkRoot(this.props)) {
+      const rest: HTMLElementProps<'a'> = omit(this.props, omitProps);
+      return (
+        <a {...rest} {...commonProps}>
+          {children}
+        </a>
+      );
+    }
+
+    const rest: HTMLElementProps<'button'> = omit(this.props, omitProps);
+
+    // NOTE(JeT): The button's `type` will be overridden if it is in the passed-in props
     return (
-      <Root
-        {...rest}
-        href={href}
-        className={cx(
-          baseStyle,
-          buttonSizes[size],
-          buttonVariants[variant],
-          className,
-        )}
-        disabled={disabled}
-        aria-disabled={disabled}
-      >
+      <button type="button" {...rest} {...commonProps}>
```

Comment:
Could we have: `const { type = 'button' } = this.props` when we first desctructure?

Reply:
I did it here because links/custom components might not have a `type` field

Reply:
Buttons should not always be `type="button"`. `type="submit"` is a good example where we might not want to explicitly set this to "button".

Reply:
Order matters with spread props. I tried to call this out in the comment right above this line, but if a consumer passes in a `type` prop, it will be part of `rest`, and will override my `type="button"` here.

Reply:
I'll add a test to make sure `type` is respected, if that helps :)

Reply:
🤦‍♂ Right, my bad! Alternatively, could we not set `type: 'button'` in default props?

Reply:
That's how it was before, actually. It was probably fine that way, but it would still be passing along `type="button"` for `<a>` tags, which we probably don't want? I found out yesterday that `type` is a valid property on `<a>` tags, meant to be used to hint to the browser the MIME type of the linked URI, so we probably don't want to set it to `"button"`, although I guess it most likely wouldn't hurt anything.

Reply:
Good point – type does have a very different meaning for `<a>` tags. On a tangential note, should we be setting `role="button"` when rendering as `<a>`?